src/Operations.jl

```diff
@@ -852,47 +853,54 @@ function download_artifacts(ctx::Context;
-                        () -> begin
-                            try
-                                download_states[name].state = :running
-                                ret()
-                                if !fancyprint
-                                    rname = rpad(name, longest_name_length)
-                                    @lock print_lock printpkgstyle(io, :Downloaded, "artifact $rname $(MiniProgressBars.pkg_format_bytes(bar.max; sigdigits=1))")
-                                end
-                            catch
-                                download_states[name].state = :failed
-                                rethrow()
-                            else
-                                download_states[name].state = :done
+    all_collected_artifacts = reduce(vcat, map(pkg_root -> collect_artifacts(pkg_root; platform), pkg_roots))
+    used_artifact_tomls = Set{String}(map(first, all_collected_artifacts))
+    longest_name_length = maximum(all_collected_artifacts; init=0) do (artifacts_toml, artifacts)
+        maximum(textwidth, keys(artifacts); init=0)
+    end
+    for (artifacts_toml, artifacts) in all_collected_artifacts
+        # For each Artifacts.toml, install each artifact we've collected from it
+        for name in keys(artifacts)
+            local rname = rpad(name, longest_name_length)
+            local hash = SHA1(artifacts[name]["git-tree-sha1"])
+            local bar = MiniProgressBar(;header=rname, main=false, indent=2, color = Base.info_color(), mode=:data, always_reprint=true)
+            local dstate = DownloadState(:ready, "", time_ns(), Base.ReentrantLock(), bar)
+            function progress(total, current; status="")
+                local t = time_ns()
+                if isempty(status)
+                    dstate.bar.max = total
+                    dstate.bar.current = current
+                end
+                lock(dstate.status_lock) do
+                    dstate.status = status
+                    dstate.status_update_time = t
+                end
+            end
+            # returns a string if exists, or function that downloads the artifact if not
+            local ret = ensure_artifact_installed(name, artifacts[name], artifacts_toml;
+                                        verbose, quiet_download=!(usable_io(io)), io, progress)
+            if ret isa Function
+                download_states[hash] = dstate
+                download_jobs[hash] =
+                    () -> begin
+                        try
+                            dstate.state = :running
+                            ret()
+                            if !fancyprint
+                                @lock print_lock printpkgstyle(io, :Downloaded, "artifact $rname $(MiniProgressBars.pkg_format_bytes(dstate.bar.max; sigdigits=1))")
                             end
+                        catch
+                            dstate.state = :failed
+                            rethrow()
```

Comment:
Should we set progress status to `failed` here and keep failed jobs around in the list until all complete/failed?

Reply:
I tried this but got strange formatting when many artifacts failed to install.

```
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
  Qt6Wayland               failed
  Gettext                  failed
  Xorg_xcb_util_wm         failed
  Libuuid                  failed
  EarCut                   failed
  Xorg_xcb_util            failed
  Vulkan_Loader            failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
  Qt6Wayland               failed
  Gettext                  failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
```

Reply:
Just need to change the counting of rows to include to include :failed

Reply:
Is it an issue if more rows than can fit in the terminal are printed?